From 33274438848815e1af089a07b80f72207a0db31f Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:07:29 -0400
Subject: [usb] uas: fix deadlocky memory allocations

Message-id: <1410887463-20674-378-git-send-email-dzickus@redhat.com>
Patchwork-id: 95586
O-Subject: [RHEL7 PATCH 377/591] uas: fix deadlocky memory allocations
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 94d72f008909610710bb1841d665eeeb010a0be1
Author: Oliver Neukum <oneukum@suse.de>
Date:   Fri Mar 28 11:25:50 2014 +0100

    uas: fix deadlocky memory allocations

    There are also two allocations with GFP_KERNEL in the pre-/post_reset
    code paths. That is no good because that is a part of the SCSI error handler.

    Signed-off-by: Oliver Neukum <oliver@neukum.org>
    Reviewed-by: Hans de Goede <hdegoede@redhat.com>
    Acked-by: Hans de Goede <hdegoede@redhat.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/storage/uas.c b/drivers/usb/storage/uas.c
index fcab9b7..511b229 100644
--- a/drivers/usb/storage/uas.c
+++ b/drivers/usb/storage/uas.c
@@ -1030,7 +1030,7 @@ static int uas_configure_endpoints(struct uas_dev_info *devinfo)
   devinfo->use_streams = 0;
  } else {
   devinfo->qdepth = usb_alloc_streams(devinfo->intf, eps + 1,
-          3, 256, GFP_KERNEL);
+          3, 256, GFP_NOIO);
   if (devinfo->qdepth < 0)
    return devinfo->qdepth;
   devinfo->use_streams = 1;
@@ -1047,7 +1047,7 @@ static void uas_free_streams(struct uas_dev_info *devinfo)
  eps[0] = usb_pipe_endpoint(udev, devinfo->status_pipe);
  eps[1] = usb_pipe_endpoint(udev, devinfo->data_in_pipe);
  eps[2] = usb_pipe_endpoint(udev, devinfo->data_out_pipe);
- usb_free_streams(devinfo->intf, eps, 3, GFP_KERNEL);
+ usb_free_streams(devinfo->intf, eps, 3, GFP_NOIO);
 }
 
 static int uas_probe(struct usb_interface *intf, const struct usb_device_id *id)
-- 
1.7.1